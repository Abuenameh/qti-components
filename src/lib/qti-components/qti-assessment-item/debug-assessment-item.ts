import { html, LitElement } from 'lit';

import { customElement, state } from 'lit/decorators.js';
import { consume, ContextConsumer } from '@lit-labs/context';
import { ItemContext, itemContext } from './qti-assessment-item.context';

@customElement('debug-assessment-item')
export class DebugAssessmentItem extends LitElement {
  // @consume({ context: itemContext, subscribe: true })
  // @state()
  // private _context: Item;
  // connectedCallback(): void {
  //   super.connectedCallback();
  //   // new ContextConsumer(
  //   //   this,
  //   //   itemContext,
  //   //   e => this.dispatchEvent(new CustomEvent('context-changed', { bubbles: true, composed: true, detail: e })),
  //   //   true
  //   // );
  // }
  // render() {
  //   return html` <pre>${JSON.stringify(this._context.variables, null, 2)}</pre> `;
  // }
}
