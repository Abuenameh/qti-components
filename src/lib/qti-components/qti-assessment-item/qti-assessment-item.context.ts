import { createContext } from '@lit-labs/context';
import { VariableDeclaration } from '../qti-utilities/Variables';

export interface ItemContext {
  href?: string;
  identifier: string;
  variables: ReadonlyArray<VariableDeclaration<string | string[]>>;
}

export const itemContext = createContext<ItemContext>('item');
