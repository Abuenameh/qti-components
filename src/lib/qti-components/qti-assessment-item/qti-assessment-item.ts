import { customElement, property } from 'lit/decorators.js';
import { html, LitElement } from 'lit';
import { OutcomeVariable } from '../qti-utilities/Variables';
import { ResponseVariable } from '../qti-utilities/Variables';
import { watch } from '../../decorators/watch';
import type { ResponseInteraction } from '../qti-utilities/ExpressionResult';
import type { Interaction } from '../qti-interaction/internal/interaction/interaction';
import type { InteractionChangedDetails, OutcomeChangedDetails } from '../qti-utilities/EventTypes';
import type { QtiFeedback } from '../qti-feedback/qti-feedback';
import type { QtiResponseProcessing } from '../qti-responseprocessing';
import type { VariableDeclaration } from '../qti-utilities/Variables';
import type QtiRegisterVariable from '../qti-utilities/events/qti-register-variable';

/**
 * @summary The qti-assessment-item element contains all the other QTI 3 item structures.
 * @documentation https://www.imsglobal.org/spec/qti/v3p0/impl#h.dltnnj87l0yj
 * @status stable
 * @since 4.0
 *
 * @dependency qti-feedback
 * @dependency qti-responseprocessing
 *
 * @slot - The default slot where all the other QTI 3 item structures go.
 *
 * @event qti-interaction-changed - Emitted when an interaction is changed.
 * @event qti-outcome-changed - Emitted when an outcome has changed.
 * @event qti-response-processing - Emitted when response-processing is called.
 *
 */
@customElement('qti-assessment-item')
export class QtiAssessmentItem extends LitElement {
  public variables: VariableDeclaration<string | string[]>[] = []; // made public for tests to access
  private feedbackElements: QtiFeedback[] = [];
  private interactionElements: Interaction[] = [];

  @property({ type: Boolean }) disabled: boolean;
  @property({ type: Boolean }) readonly: boolean;

  @property({ type: String }) title: string;
  @property({ type: String }) identifier: string;

  private _numAttempts = 0;

  override render() {
    return html`<slot></slot>`;
  }

  @watch('disabled', { waitUntilFirstUpdate: true })
  _handleDisabledChange = (_: boolean, disabled: boolean) => {
    this.interactionElements.forEach(ch => (ch.disabled = disabled));
  };

  @watch('readonly', { waitUntilFirstUpdate: true })
  _handleReadonlyChange = (_: boolean, readonly: boolean) =>
    this.interactionElements.forEach(ch => (ch.readonly = readonly));

  constructor() {
    super();
    // listen for events.
    this.addEventListener('qti-register-variable', this._registerVariable);
    this.addEventListener('qti-register-feedback', this.registerFeedbackElement);
    this.addEventListener('qti-register-interaction', this.registerInteractionElement);
    this.addEventListener('qti-outcome-changed', this.outcomeChanged);
    this.addEventListener('qti-interaction-response', this.interactionResponse);
  }

  firstUpdated(val): void {
    super.firstUpdated(val);
    this.dispatchEvent(
      new CustomEvent<{ identifier: string }>('qti-item-connected', {
        bubbles: true,
        composed: true,
        detail: this
      })
    );
  }

  public override disconnectedCallback() {
    // remove events listeners.
    this.removeEventListener('qti-register-variable', this._registerVariable);
    this.removeEventListener('qti-register-feedback', this.registerFeedbackElement);
    this.removeEventListener('qti-register-interaction', this.registerInteractionElement);
    this.removeEventListener('qti-outcome-changed', this.outcomeChanged);
    this.removeEventListener('qti-interaction-response', this.interactionResponse);
  }

  private _registerVariable(event: QtiRegisterVariable) {
    this.variables.push(event.detail.variable);
  }

  public showCorrectResponse() {
    const responseVariables = this.variables.filter(
      (vari: ResponseVariable | OutcomeVariable) => 'correctResponse' in vari && vari.correctResponse
    ) as ResponseVariable[];
    this.responses = responseVariables.map(cr => {
      return {
        responseIdentifier: cr.identifier,
        response: cr.correctResponse
      };
    });
  }

  public processResponse(): boolean {
    const responseProcessor = this.querySelector('qti-response-processing') as unknown as QtiResponseProcessing;
    if (!responseProcessor) {
      console.info('Client side response processing template not available');
      return false;
    }

    if (!responseProcessor.process) {
      console.info('Client side response webcomponents not available');
      return false;
    }

    responseProcessor.process();
    this._numAttempts++;
    this.dispatchEvent(new CustomEvent('qti-response-processing'));
    return true;
  }

  // check set response back for interactions
  // We can set the responses as a string or as an object in ResponseFormat
  set responses(myResponses: ResponseInteraction[]) {
    if (myResponses) {
      for (const response of myResponses) {
        const interaction: Interaction | undefined = this.interactionElements.find(
          i => i.getAttribute('response-identifier') === response.responseIdentifier
        );
        // If there is a responseVariable, set the value back into the responseVariable
        const responseVariable = this.getResponse(response.responseIdentifier);
        if (responseVariable) {
          responseVariable.value = response.response;
        }
        // Set the response in the interaction
        if (interaction) {
          interaction.response = response.response;
        }
      }
    }
  }

  resetInteractions() {
    this.interactionElements.forEach(interactionElement => interactionElement.reset());
  }

  // check all interactions contain valid responses
  public validateResponses(): boolean {
    let result = true;
    this.interactionElements.forEach(interactionElement => {
      if (!interactionElement.validate()) {
        result = false;
      }
    });
    return result;
  }

  public getVariable(identifier: string): VariableDeclaration<string | string[] | undefined> {
    switch (identifier) {
      case 'numAttempts':
        return {
          identifier: 'numAttempts',
          cardinality: 'single',
          baseType: 'integer',
          value: this._numAttempts.toString()
        };
        break;

      default:
        {
          const variable = this.variables.find(vr => vr.identifier === identifier);
          if (!variable) {
            console.warn(`Variable with identifier ${identifier} was not found`);
            return null;
          }
          return variable;
        }
        break;
    }
  }

  public getResponse(identifier: string): ResponseVariable | null {
    const variable = this.variables.find(vr => vr.identifier === identifier);
    const responseVariable = variable instanceof ResponseVariable ? variable : null;
    return responseVariable;
  }

  public getOutcome(identifier: string): OutcomeVariable | null {
    const variable = this.variables.find(vr => vr.identifier === identifier);
    if (variable instanceof OutcomeVariable) {
      return variable;
    }
    return null;
  }

  private registerFeedbackElement(e: CustomEvent<QtiFeedback>) {
    e.stopPropagation();
    this.feedbackElements.push(e.detail);
  }

  private registerInteractionElement(e: CustomEvent<null>) {
    e.stopPropagation();
    this.interactionElements.push(e.target as Interaction);
  }

  // interaction has fired an event to save response.
  private interactionResponse(event: CustomEvent<InteractionChangedDetails>) {
    const detail = event.detail;
    // Maybe we do not have a responseVariable declared, no problem should work without
    const responseVariable = this.getResponse(detail.responseIdentifier);
    if (responseVariable) {
      responseVariable.value = detail.response;
    }
    event.stopImmediatePropagation();

    const identifier = this.getAttribute('identifier');
    if (!identifier) {
      console.warn(`qti-assessment-item has no identifier specified`);
    }

    this.dispatchEvent(
      new CustomEvent<InteractionChangedDetails>('qti-interaction-changed', {
        bubbles: true,
        composed: true,
        detail: {
          item: identifier,
          responseIdentifier: detail.responseIdentifier,
          response: detail.response
        }
      })
    );
  }

  // checks the attributes of all feedback elements and shows/hides as appropriate.
  private outcomeChanged(event: CustomEvent<OutcomeChangedDetails>) {
    this.showFeedback(event);
  }

  private showFeedback(event: CustomEvent<OutcomeChangedDetails>) {
    this.feedbackElements.forEach(fe => {
      fe.checkShowFeedback(event.detail.outcomeIdentifier); // , event.detail.value);
    });
  }

  public setOutcomeValue(identifier: string, value: string) {
    let outcomeIdentifier: VariableDeclaration<any>;
    switch (identifier) {
      // https://qti-components.citolab.nl/?path=/docs/qti-tutorial--docs#listing-12-using-numattempts-to-set-the-completion-status
      // PK: let's fake a completionStatus variable for adaptive items
      case 'completionStatus':
        outcomeIdentifier = {
          identifier: 'completionStatus',
          cardinality: 'single',
          baseType: 'string',
          value: value
        };
        break;
      default:
        {
          outcomeIdentifier = this.getOutcome(identifier);
          if (!outcomeIdentifier) {
            console.warn(`Can not set qti-outcome-identifier: ${identifier}, it is not available`);
            return;
          }

          if (outcomeIdentifier.cardinality === 'single') {
            outcomeIdentifier.value = value;
          } else {
            (outcomeIdentifier.value as any[]).push(value);
          }
        }
        break;
    }
    this.dispatchEvent(
      new CustomEvent<OutcomeChangedDetails>('qti-outcome-changed', {
        bubbles: true,
        composed: true,
        detail: {
          item: this.identifier,
          outcomeIdentifier: identifier,
          value: outcomeIdentifier.value
        }
      })
    );
  }
}

declare global {
  interface HTMLElementTagNameMap {
    'qti-assessment-item': QtiAssessmentItem;
  }
}
