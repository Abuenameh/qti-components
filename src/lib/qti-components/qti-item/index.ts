export * from '../qti-assessment-item/qti-item.context';
export * from '../qti-assessment-item/qti-item.context.controller';
export * from './qti-item';
export * from './qti-item-logger';
