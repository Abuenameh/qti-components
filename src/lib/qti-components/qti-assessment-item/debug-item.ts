import { html, LitElement } from 'lit';

import { customElement, queryAssignedElements, queryAssignedNodes, state } from 'lit/decorators.js';
import { QtiAssessmentItem } from './qti-assessment-item';
import { register } from 'module';
import { VariableDeclaration } from '../qti-utilities/Variables';

@customElement('debug-item')
export class DebugItem extends LitElement {
  private _assessmentItemRef: QtiAssessmentItem | undefined;

  @state()
  private _items: Array<VariableDeclaration<string | string[]>> = [];

  @state()
  private _savedResponses: ReadonlyArray<VariableDeclaration<string | string[]>> = [];

  private _registerItem = (item: QtiAssessmentItem) => {
    this._assessmentItemRef = item;
    this._assessmentItemRef.context;
  };

  private _contextChanged = () => {
    this._items = [...(this._assessmentItemRef?.context.variables ?? [])];
  };

  private _saveResponse = () => {
    this._savedResponses = this._assessmentItemRef.context.variables;
  };

  private _loadResponse = () => {
    this._assessmentItemRef.context = { ...this._assessmentItemRef.context, variables: this._savedResponses };
  };
  render() {
    return html`
      <pre>${JSON.stringify(this._items, null, 2)}</pre>
      <button @click=${this._saveResponse}>Save</button>
      <button @click=${this._loadResponse}>Load</button>

      <slot
        @item-connected=${({ detail }) => this._registerItem(detail)}
        @context-changed=${() => this._contextChanged()}
      >
      </slot>
      <pre>${JSON.stringify(this._savedResponses, null, 2)}</pre>
    `;
  }
}

// @queryAssignedElements({ slot: 'list', flatten: true, selector: 'qti-assessment-item' })
// listItems!: Array<QtiAssessmentItem>;

// @slotchange=${(e: Event) => console.log('slotchange', e, this.listItems)}
