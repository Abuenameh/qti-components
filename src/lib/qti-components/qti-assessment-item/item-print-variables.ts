import { consume } from '@lit/context';
import { LitElement, html } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import { ItemContext, itemContext } from './qti-assessment-item.context';

@customElement('item-print-variables')
export class ItemPrintVariables extends LitElement {
  @consume({ context: itemContext, subscribe: true })
  @state()
  public itemContext?: ItemContext;

  render() {
    return html` <pre>${JSON.stringify(this.itemContext, null, 2)}</pre> `;
  }
}
