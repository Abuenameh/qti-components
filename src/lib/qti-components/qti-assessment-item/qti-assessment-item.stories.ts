import { html } from 'lit';
import { createRef, ref } from 'lit/directives/ref.js';

import { action } from '@storybook/addon-actions';
import { QtiAssessmentItem } from './qti-assessment-item';

import { StoryObj } from '@storybook/web-components';
import '../index';

type Story = StoryObj; // <Props>;

export default {
  component: 'qti-assesment-item'
};

const completionStatusTemplate = {
  render: args => {
    const assessmentItemRef = createRef<QtiAssessmentItem>();

    const processResponse = () => {
      assessmentItemRef.value?.processResponse();
    };
    return html`
      <qti-assessment-item
        @qti-outcome-changed="${e => {
          action(e.detail)();
        }}"
        @qti-interaction-changed="${e => {
          // processResponse();
          action(e.detail)();
        }}"
        identifier="blah"
        ${ref(assessmentItemRef)}
      >
        <qti-response-declaration identifier="RESPONSE" cardinality="single" base-type="identifier">
          <qti-correct-response>
            <qti-value>true</qti-value>
          </qti-correct-response>
        </qti-response-declaration>
        <qti-outcome-declaration identifier="FEEDBACK" cardinality="single" base-type="identifier">
        </qti-outcome-declaration>
        <qti-outcome-declaration identifier="SCORE" cardinality="single" base-type="float" normal-maximum="10.0">
          <qti-default-value>
            <qti-value>0</qti-value>
          </qti-default-value>
        </qti-outcome-declaration>
        <qti-outcome-declaration identifier="MAXSCORE" cardinality="single" base-type="float">
          <qti-default-value>
            <qti-value>10.0</qti-value>
          </qti-default-value>
        </qti-outcome-declaration>
        <qti-item-body>
          <qti-choice-interaction response-identifier="RESPONSE" shuffle="false" max-choices="1">
            <qti-prompt>Sigmund Freud and Carl Jung both belong to the psychoanalytic school of psychology.</qti-prompt>
            <qti-simple-choice identifier="true" fixed="true"
              >True<qti-feedback-inline outcome-identifier="FEEDBACK" identifier="true" show-hide="show"
                >That's correct</qti-feedback-inline
              ></qti-simple-choice
            >
            <qti-simple-choice identifier="false" fixed="true"
              >False<qti-feedback-inline outcome-identifier="FEEDBACK" identifier="false" show-hide="show"
                >That's not correct</qti-feedback-inline
              >
            </qti-simple-choice>
          </qti-choice-interaction>
          <qti-end-attempt-interaction title="end"></qti-end-attempt-interaction>
        </qti-item-body>
        <qti-response-processing>
          <qti-set-outcome-value identifier="FEEDBACK">
            <qti-variable identifier="RESPONSE"></qti-variable>
          </qti-set-outcome-value>
          <qti-response-condition>
            <qti-response-if>
              <qti-match>
                <qti-variable identifier="RESPONSE"></qti-variable>
                <qti-correct identifier="RESPONSE"></qti-correct>
              </qti-match>
              <qti-set-outcome-value identifier="SCORE">
                <qti-variable identifier="MAXSCORE"></qti-variable>
              </qti-set-outcome-value>
            </qti-response-if>
          </qti-response-condition>
        </qti-response-processing>
      </qti-assessment-item>
    `;
  }
};

export const InlineFeedback: Story = {
  render: completionStatusTemplate.render,
  args: {},
  play: async ({ canvasElement }) => {
    // const canvas = within(canvasElement);
    // const choiceElement = canvas.getByTestId('choice');
    // const isFeedbackVisible = canvas.getByTestId('feedback').shadowRoot.querySelector('slot');
    // await userEvent.click(choiceElement);
    // await new Promise(resolve => setTimeout(resolve, 500));
    // expect(isFeedbackVisible).toHaveClass('on');
    // await userEvent.click(choiceElement);
    // await new Promise(resolve => setTimeout(resolve, 500));
    // expect(isFeedbackVisible).toHaveClass('off');
  }
};
