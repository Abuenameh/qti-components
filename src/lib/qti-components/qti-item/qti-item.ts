import { Signal, SignalWatcher, computed, html, signal } from '@lit-labs/preact-signals';
import { LitElement } from 'lit';
import { customElement } from 'lit/decorators.js';
import { OutcomeChangedDetails } from '../internal/event-types';
import { QtiAssessmentItem } from '../qti-assessment-item/qti-assessment-item';

@customElement('qti-item')
export class QtiItem extends SignalWatcher(LitElement) {
  private batchedOutcomeEvents: OutcomeChangedDetails[] = [];
  private debounceTimeout: NodeJS.Timeout;
  debounceResponseDelay: number = 500;

  public context: Signal<
    {
      identifier: string;
      variables: Signal<{ identifier: string; value: Readonly<string | string[]> }[]>;
    }[]
  > = signal([]);

  constructor() {
    super();
    this.addEventListener('qti-assessment-item-connected', (e: CustomEvent<Signal>) => {
      const qtiAssessmentItem = e.composedPath().find(el => el instanceof QtiAssessmentItem) as QtiAssessmentItem;
      this.updateContext(qtiAssessmentItem);
    });
  }

  updateContext(qtiAssessmentItem: QtiAssessmentItem) {
    const existing = this.context.value.find(v => v.identifier === qtiAssessmentItem.identifier);

    const variables = computed(() =>
      qtiAssessmentItem.context.value
        .filter(v => v.identifier === 'RESPONSE')
        .map(v => ({ identifier: v.identifier, value: v.value }))
    );

    if (existing) {
      qtiAssessmentItem.context.value = qtiAssessmentItem.context.value.map(v => {
        const existingVariable = existing.variables.value.find(e => e.identifier === v.identifier);
        return existingVariable ? { ...v, value: existingVariable.value } : v;
      });
    }

    this.context.value = [
      ...this.context.peek().filter(v => v.identifier !== qtiAssessmentItem.identifier),
      { identifier: qtiAssessmentItem.identifier, variables }
    ];
  }

  render() {
    return html`
      <h1>QTI Item</h1>
      <pre>${computed(() => JSON.stringify(this.context, null, 4))}</pre>
      <slot></slot>
    `;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    'qti-item': QtiItem;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    'qti-item': QtiItem;
  }
}
