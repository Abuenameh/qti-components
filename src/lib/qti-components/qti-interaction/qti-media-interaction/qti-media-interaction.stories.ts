import { html } from 'lit';
import { action } from '@storybook/addon-actions';
import { useRef, virtual } from 'haunted';
import '../../../qti-item';

import xml from './qti-media-interaction.xml?raw';
import { QtiAssessmentItem } from '../../qti-assessment-item/qti-assessment-item';

export default {
  component: 'qti-media-interaction',
  decorators: [story => html`${virtual(story)()}`],
  argTypes: {
    autostart: { type: 'boolean', description: '' },
    minPlays: { type: 'number', description: '' },
    maxPlays: { type: 'number', description: '' },
    loop: { type: 'boolean', description: '' }
    // coords: { type: 'object', description: '' },
  }
  // args: { autostart: false, minPlays: 0, maxPlays: 0, loop: false },
};

export const Default = {
  render: args => html`
    <qti-media-interaction
      autostart=${args.autostart}
      loop=${args.loop}
      min-plays=${args.minPlays}
      max-plays=${args.maxPlays}
      response-identifier="RESPONSE"
      @qti-interaction-response="${e => {
        action(e);
      }}"
      .response=${args.response}
      ?disabled=${args.disabled}
      ?readonly=${args.readonly}
    >
      <qti-prompt>Play this video.</qti-prompt>
      <video width="320" height="240">
        <source src="qti-media-interaction/earth.mp4" type="video/mp4" />
        Your browser does not support the video tag
      </video>
    </qti-media-interaction>
  `,

  args: {}
};

export const Item = {
  render: () => {
    const qtiItemRef = useRef<QtiAssessmentItem>(null);
    return html`<qti-item
        @qti-outcome-changed=${action(`qti-outcome-changed`)}
        @qti-interaction-changed=${action(`qti-interaction-changed`)}
        @qti-item-connected=${({ detail }) => (qtiItemRef.current = detail)}
        xml=${xml}
      ></qti-item>
      <button @click=${() => qtiItemRef.current.processResponse()}>PROCESS</button>`;
  }
};
