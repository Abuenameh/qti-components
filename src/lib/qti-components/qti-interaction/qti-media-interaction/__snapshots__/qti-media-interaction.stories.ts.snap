// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/qti-components/qti-interaction/qti-media-interaction Default smoke-test 1`] = `
<div id="root-inner">
  <qti-media-interaction autostart="false"
                         response-identifier="RESPONSE"
                         loop
                         min-plays
                         max-plays
  >
    <qti-prompt>
      Play this video.
    </qti-prompt>
    <video width="320"
           height="240"
           controls
    >
      <source src="qti-media-interaction/earth.mp4"
              type="video/mp4"
      >
      Your browser does not support the video tag
    </video>
  </qti-media-interaction>
</div>
`;
